src/main/java/net/snowflake/client/jdbc/SnowflakeSQLLoggedException.java

```diff
@@ -0,0 +1,238 @@
+{
+  /*
+  OOB telemetry instance
+   */
+  private TelemetryService oobInstance = TelemetryService.getInstance();
+
+  /*
+  IB telemtry instance
+   */
+  private Telemetry ibInstance = null;
+
+  private final static ObjectMapper mapper =
+          ObjectMapperFactory.getObjectMapper();
+
+  /**
+   * Function to create a TelemetryEvent log from the JSONObject and exception and send it via OOB telemetry
+   *
+   * @param value JSONnode containing relevant information specific to the exception constructor that
+   *              should be included in the telemetry data, such as sqlState or vendorCode
+   * @param ex    The exception being thrown
+   */
+  private void sendOutOfBandTelemetryMessage(JSONObject value, SnowflakeSQLLoggedException ex)
+  {
+    TelemetryEvent.LogBuilder logBuilder = new TelemetryEvent.LogBuilder();
+    StringWriter sw = new StringWriter();
+    PrintWriter pw = new PrintWriter(sw);
+    ex.printStackTrace(pw);
+    String stackTrace = sw.toString();
+    value.put("Stacktrace", stackTrace);
+    TelemetryEvent log = logBuilder
+        .withName("Exception: " + ex.getMessage())
+        .withValue(value)
+        .build();
+    oobInstance.report(log);
+  }
+
+  /**
+   * Function to create a TelemetryClient log and send it via in-band telemetry
+   *
+   * @param value ObjectNode containing information specific to the exception constructor that should be included in
+   *              the telemetry log, such as SQLState or reason for the error
+   * @param ex The exception being thrown
+   * @return true if in-band telemetry log sent successfully or false if it did not
+   */
+  private boolean sendInBandTelemetryMessage(ObjectNode value, SnowflakeSQLLoggedException ex) {
+    StringWriter sw = new StringWriter();
+    PrintWriter pw = new PrintWriter(sw);
+    ex.printStackTrace(pw);
+    String stackTrace = sw.toString();
+    value.put("Stacktrace", stackTrace);
+    ibInstance.addLogToBatch(new TelemetryData(value, System.currentTimeMillis()));
+    Future<Boolean> success = ibInstance.sendBatchAsync();
+    try
+    {
+      // wait for 10 seconds at most before giving up
+      return success.get(10, TimeUnit.SECONDS);
+    }
+    catch (Exception e)
+    {
+      success.cancel(true);
+      return false;
+    }
+  }
+
+  /**
+   * Function to construct log data based on possible exception inputs and send data through in-band telemetry, or oob
+   * if in-band does not work
+   * @param queryId query ID if exists
+   * @param reason reason for the exception
+   * @param SQLState SQLState
+   * @param vendorCode vendor code
+   * @param errorCode error code
+   * @param session session object, which is needed to send in-band telemetry but not oob. Might be null, in which case
+   *                oob is used.
+   */
+  public void sendTelemetryData(String queryId, String reason, String SQLState, int vendorCode, ErrorCode errorCode, SFSession session)
```

Comment:
This function gets called every time after super(). If in-band message is not sent, oob message is sent instead.